app/src/main/java/com/example/stretchy/features/executetraining/ui/ExecuteTrainingViewModel.kt

```diff
@@ -52,95 +62,241 @@ class ExecuteTrainingViewModel(val repository: Repository, val trainingId: Long)
         }
         _uiState.value = _uiState.value.copy(isLoading = true)
         viewModelScope.launch {
-            val trainingWithActivities = repository.getTrainingWithActivitiesById(trainingId)
-            getExercisesWithBreak(trainingWithActivities.activities).let { activitiesWithBreaks ->
-                activitiesWithBreaks.forEachIndexed { index, activity ->
-                    timer.setDuration(activity.duration)
-                    timer.flow.takeWhile { it >= 0 }.collect { currentSeconds ->
-                        when (activity.activityType) {
-                            ActivityType.STRETCH -> {
-                                handleStretch(
-                                    activity,
-                                    activitiesWithBreaks.getOrNull(index + 2)?.name,
-                                    currentSeconds,
-                                    index == 0
-                                )
-                            }
-                            ActivityType.BREAK -> {
-                                handleBreak(
-                                    activity,
-                                    activitiesWithBreaks.getOrNull(index + 1)?.name,
-                                    currentSeconds
-                                )
-                            }
-                            else -> {}
-                        }
+            trainingWithActivities = repository.getTrainingWithActivitiesById(trainingId)
+            initializeDisplayableList()
+
+            trainingWithActivities.activities.let { activitiesWithBreaks ->
+                while (true) {
+                    val currentActivity = activitiesWithBreaks[index]
+
+                    presetCurrentExercise(currentActivity)
+                    countDownTimeCurrentActivity(currentActivity)
+                    afterActivityEnds(currentActivity)
```

Comment:
also here. The function name describes `when` it happens -> after activity ends. Which brings no value. I cannot read the code as a whole without going into the function details to see what it does.
It says **nothing** about what the function actually `do`.

Reply:
What do you think about naming it ```handleActicityEnd```?

Reply:
I would rather name it more about something to do with pages, if you look at the code its about page chainging and indexes. So that would be more precise